ocaml/xapi/xapi_vm_group_helpers.ml

```diff
@@ -0,0 +1,208 @@
+(*
+ * Copyright (C) 2024 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "xapi_vm_group_helpers" end)
+
+open D
+
+let check_breach_on_vm_anti_affinity_rules ~__context ~group =
+  Db.VM_group.get_VMs ~__context ~self:group
+  |> List.filter_map (fun vm ->
+         let vm_rec = Db.VM.get_record ~__context ~self:vm in
+         match (vm_rec.API.vM_power_state, vm_rec.API.vM_resident_on) with
+         | `Running, h when h <> Ref.null ->
+             Some h
+         | _ ->
+             None
+     )
+  |> function
+  | [] | [_] ->
+      None
+  | h :: remaining ->
+      if List.exists (fun h' -> h' <> h) remaining then
+        None
+      else
+        Some h
+
+let report_anti_affinity_alert ~__context ~group ~host =
+  let group_uuid = Db.VM_group.get_uuid ~__context ~self:group in
+  let host_uuid = Db.Host.get_uuid ~__context ~self:host in
+  let body =
+    Printf.sprintf
```

Comment:
`Printf.sprintf` is useless here.

Reply:
It is used for formatting, as the original similar implementation, I will keep it

Reply:
If no argument, `Printf.sprintf` is no needed. Or you change it to this:
```
    let body =
      Printf.sprintf
        "<body><message>%s</message><VM_group>%s</VM_group><host>%s</host></body>"
        "Breach on VM anti-affinity rules" group_uuid host_uuid
    in

Reply:
I will remove  `Printf.sprintf `from code